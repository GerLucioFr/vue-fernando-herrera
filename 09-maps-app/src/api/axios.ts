import axios from 'axios';

export type CustomParamValue =
  | string
  | string[]
  | number
  | number[]
  | boolean
  | null
  | undefined;

interface CustomParams {
  [key: string]: CustomParamValue;
}

export const axiosInstance = (
  baseURL: string,
  params?: CustomParams,
  headers?: CustomParams
) => {
  return axios.create({
    baseURL,
    params,
    headers,
  });
};
