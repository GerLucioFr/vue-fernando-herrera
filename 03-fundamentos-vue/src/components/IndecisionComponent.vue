<template>
  <div class="indecision-container">
    <img
      :src="
        questionAnswer.image ||
        'https://w.wallhaven.cc/full/96/wallhaven-966re1.jpg'
      "
      alt="background"
    />
    <div class="indecision-form">
      <input
        type="text"
        v-model="question"
        :disabled="isLoading"
        placeholder="Ask me anything?"
      />
      <p>Remember to use a question mark at the end</p>

      <div class="question">
        <h2>{{ question || 'Your question goes here' }}</h2>

        <h1 v-if="isLoading">Loading your answer :)</h1>
        <h1 v-if="!isLoading && questionAnswer.answer">
          {{ questionAnswer.answer }}
        </h1>
      </div>
    </div>
  </div>
</template>
<script>
import axios from 'axios';
export default {
  name: 'IndecisionComponent',
  data() {
    return {
      question: '',
      isLoading: false,
      questionAnswer: {
        answer: '',
        forced: false,
        image: '',
      },
    };
  },
  watch: {
    question(value) {
      if (!value.includes('?')) {
        this.resetAnswer();
        return;
      }
      this.getAnswer();
    },
  },
  methods: {
    async getAnswer() {
      const baseURL = 'https://yesno.wtf/api';
      const axiosInstance = axios.create({
        baseURL,
      });

      this.isLoading = true;
      try {
        const { data } = await axiosInstance.get();

        this.questionAnswer = data;
        this.isLoading = false;
      } catch (error) {
        this.isLoading = false;
        console.error(error);
      }
    },
    resetAnswer() {
      this.questionAnswer = {
        answer: '',
        forced: false,
        image: '',
      };
    },
  },
};
</script>
<style scoped>
.indecision-container {
  box-sizing: border-box;
  position: relative;
  width: 100%;
  height: 100vh;
}

img {
  position: absolute;
  top: 0;
  left: 0;

  width: 100%;
  height: 100%;
  object-fit: cover;

  opacity: 0.5;

  z-index: 1;
}

.indecision-form {
  position: relative;

  margin: 32px 0;

  z-index: 2;
}

input[type='text'] {
  padding: 8px;
  border-radius: 4px;
}

.question {
  margin-top: 64px;
}
</style>
