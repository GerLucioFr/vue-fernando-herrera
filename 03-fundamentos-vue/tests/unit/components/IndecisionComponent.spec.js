import { shallowMount } from '@vue/test-utils';
import IndecisionComponent from '@/components/IndecisionComponent';
import axios from 'axios';

describe('Indecision Component Test', () => {
  let wrapper;
  let input;
  let getAsnwerSpy;

  beforeEach(() => {
    wrapper = shallowMount(IndecisionComponent);
    input = wrapper.find('#question');
    getAsnwerSpy = jest.spyOn(wrapper.vm, 'getAnswer');

    jest.clearAllMocks();
  });

  it('should to match with snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should not to trigger any event on input (no question mark)', async () => {
    await input.setValue('Hello world');
    expect(getAsnwerSpy).not.toHaveBeenCalled();
  });

  it('should to trigger an event on question mark input', async () => {
    await input.setValue('Are your ready?');
    expect(getAsnwerSpy).toHaveBeenCalledTimes(1);
  });

  it('should to make http request to get data', async () => {
    await wrapper.vm.getAnswer();
    expect(wrapper.vm.questionAnswer?.image).toBeTruthy();
  });

  it('should to fail http request', async () => {
    const mockAxiosInstance = {
      get: jest.fn(),
    };
    jest.spyOn(axios, 'create').mockReturnValueOnce(mockAxiosInstance);

    mockAxiosInstance.get.mockRejectedValueOnce(
      new Error('Request failed with status code 500')
    );

    await wrapper.vm.getAnswer.call({
      axiosInstance: mockAxiosInstance,
    });
    expect(wrapper.vm.questionAnswer?.image).toBeFalsy();

  });
});
