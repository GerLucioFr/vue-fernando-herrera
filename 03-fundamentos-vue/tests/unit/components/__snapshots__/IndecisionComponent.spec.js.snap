// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Indecision Component Test should to match with snapshot 1`] = `
<div class="indecision-container"><img src="https://w.wallhaven.cc/full/96/wallhaven-966re1.jpg" alt="background">
  <div class="indecision-form"><input type="text" id="question" placeholder="Ask me anything?">
    <p>Remember to use a question mark at the end</p>
    <div class="question">
      <h2>Your question goes here</h2>
      <!--v-if-->
      <!--v-if-->
    </div>
  </div>
</div>
`;
